### PR TITLE
Statusbar color issue fix 

### PR DESCRIPTION
**Pull request by sam33rdhakal**
Updates
* Instead of getting the specific color we pull the color by attribute defined in active theme so that it is dynamic and applies correct color.
* Marks the fragments `AloofFragment` if they are not suppose to modify parent activity's theme

### CODE REVIEW

File: base/src/main/java/edu/artic/base/utils/ActivityExtensions.kt [hunk 1]
```diff
@@ -83,4 +85,22 @@ val Activity.statusBarHeight: Int
...
+/**
+ * Takes the array of color attributes ids and return array of colors.
+ */
+fun Context.getThemeColors(colors: IntArray): Array<ColorStateList?> {
+    val found: Array<ColorStateList?>
+    var typedArray: TypedArray? = null
+    try {
+        typedArray = this.obtainStyledAttributes(colors)
+        found = arrayOfNulls(colors.size)
+        for (i in 0 until colors.size) {
```
Comment (agrosner) [inline base/src/main/java/edu/artic/base/utils/ActivityExtensions.kt]: could you do `found = (0 until colors.size).map { typedArray.getColorStateList(it) }.toTypedArray()` ?
Also why is this necessary? Why not just set the status bar color to the kind we want?

File: ui/src/main/kotlin/edu/artic/ui/BaseFragment.kt [hunk 2]
```diff
@@ -123,4 +129,9 @@ abstract class BaseFragment : Fragment() {
...
 
 
-}
\ No newline at end of file
+}
+
+/**
+ * Marker interface to mark those fragments which shouldn't modify the parent activity theme.
+ */
+interface AloofFragment
```
Comment (agrosner) [inline ui/src/main/kotlin/edu/artic/ui/BaseFragment.kt]: Give it a better name. or better yet, just use an open `val` that they can override to control this behavior.

Comment (Cliabhach) [reply ui/src/main/kotlin/edu/artic/ui/BaseFragment.kt]: Oh, good idea. `open val mayOverrideStatusBarColor` or something like that?

File: base/src/main/java/edu/artic/base/utils/ActivityExtensions.kt (hunk 1 above)
Comment (Cliabhach) [reply base/src/main/java/edu/artic/base/utils/ActivityExtensions.kt]: We don't know what color to use until the fragment is added.

File: ui/src/main/kotlin/edu/artic/ui/BaseFragment.kt (hunk 2 above)
Comment (agrosner) [reply ui/src/main/kotlin/edu/artic/ui/BaseFragment.kt]: yeah to follow convention in other parts of it

File: base/src/main/java/edu/artic/base/utils/ActivityExtensions.kt (hunk 1 above)
Comment (sam33rdhakal) [reply base/src/main/java/edu/artic/base/utils/ActivityExtensions.kt]: @agrosner Updated the code. @Cliabhach is right for the second point.

File: ui/src/main/kotlin/edu/artic/ui/BaseFragment.kt (hunk 2 above)
Comment (sam33rdhakal) [reply ui/src/main/kotlin/edu/artic/ui/BaseFragment.kt]: Replaced the marker interface by parent class property.

File: ui/src/main/kotlin/edu/artic/ui/BaseFragment.kt [hunk 3]
```diff
@@ -93,11 +93,16 @@ abstract class BaseFragment : Fragment() {
...
-            requireActivity().setWindowFlag(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS, false)
-            requireActivity().window?.statusBarColor = Color.TRANSPARENT
-        } else {
-            requireActivity().window?.statusBarColor = ContextCompat.getColor(requireContext(), R.color.colorPrimary)
+        if (this !is AloofFragment) {
+            if (hasTransparentStatusBar()) {
+                requireActivity().setWindowFlag(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS, false)
+                requireActivity().window?.statusBarColor = Color.TRANSPARENT
+            } else {
+                val primaryDarkColor = intArrayOf(android.support.design.R.attr.colorPrimaryDark)
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply ui/src/main/kotlin/edu/artic/ui/BaseFragment.kt]: Leaving the code to use colorPrimaryDark for now. We could refactor it later in different merge.

File: base/src/main/java/edu/artic/base/utils/ActivityExtensions.kt [hunk 4]
```diff
@@ -83,4 +85,21 @@ val Activity.statusBarHeight: Int
...
+/**
+ * Takes the array of color attributes ids and return array of colors.
+ */
+fun Context.getThemeColors(colors: IntArray): Array<ColorStateList?> {
+    val found: Array<ColorStateList?>
+    var typedArray: TypedArray? = null
+    try {
+        typedArray = this.obtainStyledAttributes(colors)
+        found = (0 until colors.size)
+                .map { typedArray.getColorStateList(it) }
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply base/src/main/java/edu/artic/base/utils/ActivityExtensions.kt]: Makes sense. Updated.

File: ui/src/main/kotlin/edu/artic/ui/BaseFragment.kt [hunk 5]
```diff
@@ -93,11 +99,16 @@ abstract class BaseFragment : Fragment() {
...
-            requireActivity().window?.statusBarColor = Color.TRANSPARENT
-        } else {
-            requireActivity().window?.statusBarColor = ContextCompat.getColor(requireContext(), R.color.colorPrimary)
+        if (overrideStatusBarColor) {
+            if (hasTransparentStatusBar()) {
+                requireActivity().setWindowFlag(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS, false)
+                requireActivity().window?.statusBarColor = Color.TRANSPARENT
+            } else {
+                val primaryDarkColor = intArrayOf(android.support.design.R.attr.colorPrimaryDark)
+                requireContext().getThemeColors(primaryDarkColor)[0]?.defaultColor?.let {
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply ui/src/main/kotlin/edu/artic/ui/BaseFragment.kt]: Updated.

**Comment by Cliabhach** [comment]
Ah, but first I must rebase....
